### ISSUE TITLE
Question about compatibility implementation

### ISSUE THREAD

**Issue by n0ct** [stefvanschie/IF#155]
Hi guys. I just discovered your work and it's really amazing. I was thinking about creating a menu creation library but you developed almost everything I need.
**I'm now wondering about how to plug my item library into your framework.**

As I am not familiar with your framework, I am creating this issue to exchange with you on this subject and to think about the best way to do what I need. I hope I am not disturbing you in doing so, and if I am, please excuse me and tell me how I should go about it (where to ask such questions).

My library (called, without more imagination, ItemLib) allows plugin developers to :

- Define all the plugin's items quickly and simply using JSON files.
- Let server owners modify those items in JSON files.
- Update those item data as a developer (and override the modifications done by server owners if required).
- Translate those item names and lore in multiple languages.
- Translate any other text (like messages).
- Define all the plugin's translations quickly and simply using JSON files.
- Let server owners modify those translations in JSON files.
- Update those translations as a developer (and override the modifications done by server owners if required).
- Set and modify the player lang.
- Listen to events related to each item and triggered according to the slot in which this object is located.
- This event system simplifies the detection of player clicks because it contains an ItemClickEvent that removes the complexities of the PlayerInterractEvent, EntityDamageByEntityEvent and PlayerAnimationEvent bukkit events.
- Save the player inventory and allow it to be restored automatically depending on event conditions (which can persist even after a server restart).
- Create/Modify ItemStacks using an ItemStackWrapper (and ItemStackUtils).
- Compare/Count/remove ItemStacks using ItemStackTrait (and ItemStackUtils).
- Create custom heads based on custom textures.
- Create items based on custom 3D models / textured according to a texture pack.
- 1.12-1.17 compatibility

To be precise it is based on 3 other libs I made (JsonLib, TranslationLib and event-wrapper-lib) but that does not change anything for your understanding.

ItemLib is currently private but it will be published with an open source license in the next months.

**I would therefore like to slightly modify or extend your framework in order to meet several needs.**

These are my needs:
1 - Show ItemStack according to the player's language.
  - [In my lib] An ItemStack can be retrieved using:
     - a String "technicalName" which is an unique identifier
     - a Language (that can be retrieved from the player)
  - [How ?] Could I extend GuiItem in order to do so. Is it possible to retrieve the player that will see the GUI when working with a GuiItem ?

2 - Translate every text according to the player's language
  - [In my lib] A translated text can be retrieved using:
    - a String "key" which is the translation key
    - a Language (that can be retrieved from the player)
  => I don't even know every places where there can be a text message in your framework but since I saw you made adventure compatibility possible, maybe I could follow the same technical design to allow compatibility with my translation system. Or should I plug into adventure instead ?
  
3 - Be able to refresh/close a GUI according to an event (for example, if the GUI is related to a block and if another player destroy this block, the GUI should be closed)

Please advice

**Comment by stefvanschie** [comment]
A single gui instance is the same for all viewers. Therefore if you need to display different items or different text for different players (for example, due to their language setting), you should use multiple different gui instances. It's probably easiest if you create a gui instance for a single person. That way you can adjust this single gui instance to the player's settings. This should hopefully answer questions one and two.

Refreshing a gui can be done via `Gui#update`. Closing a gui for a player can simply be done via `Player#closeInventory`. You can simply call these methods in your event handler.

IF allows you to place panes inside the player's inventory when they have an inventory open. For merged guis (these are barrel, chest, ender chest and shulker box guis) this can simply by done by moving a pane lower done. The gui is assumed to extend across both the top inventory and bottom inventory and act as one whole. For the other guis, you can place the panes in the player inventory component by first getting this component and then adding the pane to it. If you place a pane in the bottom inventory, IF will take care of saving the player's inventory before the gui is opened and restoring it after the gui is closed again.

Hopefully this answers your questions.

**Comment by n0ct** [comment]
Thank you for your very clear answers. There is only one point I am not sure I understood:

If I understood correctly, it is not possible to use IF to modify the shortcutbar if there is no open inventory. In other words, IF can save the player's inventory, modify the shortcut bar, and then, when closing the GUI, put the saved inventory back. But all this can only be done when opening (and then closing) an inventory.

Do I have this right?

**Comment by n0ct** [comment]
Another question comes to mind:
You were telling me that IF causes the player's inventory to be reset when the GUI closes (in case the GUI temporarily overwrites their inventory).

What happens if:
1 - A player leaves the server while they have this GUI open?
2 - A player dies with this GUI open (and inventory retention is enabled/disabled on the server)?
3 - Or what if the server crashes at that moment?

Will the player's inventory be restored in all 3 cases? And in the second case, will the GUI items fall to the ground?

**Comment by stefvanschie** [comment]
Yes, that's correct, the inventory of the player can only be modified while a gui is opened since IF is intended for guis, so it essentially does nothing if there's no open gui.

The intention is that the player's inventory is always restored (and gui items are not dropped on the ground). I would need to test these specific scenarios you mentioned (and update IF if they happen to not work), but the idea is that these cases are handled. (A server crashing might be a bit troublesome, though.)

**Comment by n0ct** [comment]
Hi,

Thanks again for your help. With a developer of the association in which I am a volunteer, we started to work on this modified version of IF. Unfortunately we were quickly **blocked by a small configuration problem**.
```
Cannot resolve io.papermc:paper:1.16.3-R0.1-SNAPSHOT
Cannot resolve org.spigotmc:spigot:1.17-R0.1-SNAPSHOT
Cannot resolve io.papermc:paper:1.15.2-R0.1-SNAPSHOT
Cannot resolve org.spigotmc:spigot:1.18-R0.1-SNAPSHOT
Cannot resolve io.papermc:paper:1.16.1-R0.1-SNAPSHOT
Cannot resolve io.papermc:paper:1.14.4-R0.1-SNAPSHOT
Cannot resolve org.spigotmc:spigot:1.17.1-R0.1-SNAPSHOT
Cannot resolve org.spigotmc:spigot:1.18.2-R0.1-SNAPSHOT
Cannot resolve io.papermc:paper:1.16.4-R0.1-SNAPSHOT
```

To fix this problem:
- For missing spigot dependencies it is probably enough to run BuildTools.jar locally for each desired version. Is that right ?
- But for Paper it seems to me that it is more complex and I don't know exactly how to do it. I have the impression that I had to run the paper.jar of each version with `-Dpaperclip.install=true -jar` but some of the versions IF uses as dependencies are not available for download on the paper website, even in their archives. What should we do exactly ?

Could you please either complete your README.md or your wiki with a contributing guide explaining how to prepare the development environment. Or, simply explain it to us right here ?

**Comment by stefvanschie** [comment]
Thank you for your interest in addressing this. Yes, for the Spigot dependencies, running BuildTools for the relevant versions is all you need to do to obtain these files. For the Paper dependencies, the versions are downloaded via the API. Then you can run these JAR files to obtain the patched JAR file. These are then manually installed in the local repository via Maven with the following command.
```
mvn install:install-file -Dfile=cache/patched_1.16.4.jar -DgroupId="io.papermc" -DartifactId="paper" -Dversion="1.16.4-R0.1-SNAPSHOT" -Dpackaging="jar"
```
This is an example for version 1.16.4, but the same is used for other versions, but with different version numbers.

That should be everything to get a working environment, but I'll make sure to update the instructions in the README.md as well, so it is clearer as to how you can obtain this environment.

**Comment by stefvanschie** [comment]
This has been open for several years now with no real work on it from my part. The remaining open point was regarding hard crashes on the server (e.g., due to a power failure). Ensuring this works optimally is a significant undertaking, which is why I've not really worked on it. However, after further investigation, I do not think it is worth pursuing this.

The server itself does not guard against these hard crashes either. While it is possible for the server to be in a safe state after a hard crash, there is also a chance that it might not be. The server does not guarantee that it will still work properly. Because of that, if a hard crash occurs, the only safe, practical way to recover is to restore a backup. To that end, implementing this functionality feels relatively in vain, because a GUI that has been restored properly is not very useful if many other parts of the server do not recover correctly. Because of this, I see little reason to implement this.

Note that this is only for hard crashes. If the server is shut down, IF will restore everything properly like it has always done.
